BE - 댓글 및 대댓글 기능 구현

gomljo opened the pull request:
## Summary
댓글 및 대댓글 기능 구현
- 댓글 CRUD
- 대댓글 CRUD
- 레시피 객체의 댓글 수 증감 기능

 
## Describe your changes

- 댓글 CRUD
  - 댓글 생성
    - 댓글 생성 시 레시피의 댓글 수 증가
  - 댓글 수정
  - 댓글 삭제
    - 댓글 삭제 시 레시피의 댓글 수 감소
  - 댓글 조회
  - 레시피에 대해 작성된 댓글 목록 조회
    
- 대댓글 CRUD
  - 대댓글 생성
    - 대댓글 생성 시 레시피의 댓글 수 증가
  - 대댓글 수정
  - 대댓글 삭제
    - 대댓글 삭제 시 레시피의 댓글 수 감소
  - 대댓글 조회
  - 레시피에 대해 작성된 대댓글 목록 조회

- 레시피 객체의 댓글 수 증감 기능
  - 댓글 및 대댓글 생성 또는 삭제 시 레시피의 댓글 수도 증감되도록 구현

## Issue number and link
#101

gomljo committed:
feat: 댓글 도메인 및 repository 정의

댓글 도메인
- 댓글 내용 수정 기능 구현
- 댓글 삭제 여부 수정 기능 구현

repository
- 작성한 사람의 댓글 가져오기 기능 구현
- 레시피에 대한 댓글 리스트 가져오는 기능 구현

resolves: #101

gomljo committed:
feat: 댓글 CRUD 구현

구현 목록

- 댓글 생성 기능 구현
- 댓글 목록 조회 기능 구현
- 댓글 조회 기능 구현
- 댓글 삭제 기능 구현
- 댓글 수정 기능 구현

resolves: #101

gomljo committed:
feat: 대댓글 domain 및 repository 정의

대댓글 domain
- 대댓글 수정 기능 구현

대댓글 repository
- 대댓글 조회하기 구현
- 레시피에 작성된 대댓글 목록 조회하기 구현

resolves: #101

gomljo committed:
feat: 대댓글 CRUD 서비스 구현

구현 목록

- 대댓글 생성 기능 구현
- 대댓글 삭제 기능 구현
- 대댓글 목록 조회 기능 구현
- 대댓글 수정 기능 구현

resolves: #101

gomljo committed:
feat: 댓글, 대댓글 facade 서비스 구현

댓글과 대댓글 작성 및 삭제 시 레시피의 댓글 수에 대한 동시성 제어가 필요
따라서 댓글과 대댓글 작성 및 삭제 시 레시피와 함께 lock을 사용하도록
구현

resolves: #101

gomljo committed:
feat: 댓글 및 대댓글 요청 컨트롤러 구현

댓글 및 대댓글 CRUD 요청을 처리할 수 있는 컨트롤러 구현

resolves: #101

gomljo committed:
feat: 댓글 수 증감 기능 구현

댓글 및 대댓글 작성 및 삭제 시 댓글 수 증감이 되도록 기능 구현

resolves: #101

gomljo committed:
refactor: 댓글 및 대댓글 목록 조회 순서 변경

댓글 및 대댓글 조회 시 생성 시점을 기준으로 오름차순으로 목록을
반환하도록 변경

resolves: #101

SeungHyunLee054 reviewed (approved):
확인 완료했습니다! 고생 많으셨습니다!
setter 사용을 지양하기 위해서 entity에 메소드로 정의하는 게 좋은 방법인 것 같습니다! 
저도 그런식으로 작성하도록 하겠습니다!

gomljo committed:
feat: 레시피 검색 객체 정보 업데이트 기능 구현

레시피에 대한 댓글, 대댓글의 증감에 따라 레시피 검색 객체의 댓글 수도 증감하기 때문에 레시피 검색 객체의 정보 업데이트 기능 구현

resolves: #101

J-HyeonSeo commented on src/main/java/com/zerobase/foodlier/global/comment/facade/reply/ReplyFacade.java:
```diff
@@ -0,0 +1,48 @@
+    private final RecipeService recipeService;
+
+    @Transactional
+    @RedissonLock(group = "reply", key = "#recipeId")
```
같은 데이터를 변경하는데,  댓글과 대댓글의 group이 다릅니다.
해당 부분을 하나로 통일하면 좋을 것 같습니다!

gomljo replied:
넵! 수정했습니다!

J-HyeonSeo reviewed (approved):
댓글 및 대댓글에 대한 CRUD 확인하였습니다.
고생많으셨습니다.

gomljo committed:
chore: 불필요한 어노테이션 제거

CommentService 인터페이스에 존재하는 불필요한 어노테이션 제거

resolves: #101

gomljo committed:
refactor: 대댓글 lock 파라미터 수정

레시피의 댓글 수 증감에 대한 동시성 처리를 위해 댓글 lock 어노테이션에 정의된 group과 같은 group으로 할당

resolves: #101

thwang26 reviewed (approved):
댓글 및 대댓글 기능 코드 작성 확인하였습니다.
코드 작성하시느라 고생 많으셨습니다!
